src/api/alchemy-wallet.ts

```diff
@@ -1,3 +1,105 @@
+import { Alchemy } from './alchemy';
+import { BigNumber } from '@ethersproject/bignumber';
+export class Wallet extends EthersWallet {
+  readonly hasAlchemyObject: boolean = false;
+  readonly alchemy?: Alchemy;
+  alchemyProvider?: Provider;
 
-export class Wallet extends EthersWallet {}
+  constructor(
+    privateKey: BytesLike | ExternallyOwnedAccount | SigningKey,
+    alchemyOrProvider?: Alchemy | Provider
+  ) {
+    // If object passed in is a provider, send to super
+    let superProvider: Provider | undefined;
+    if (alchemyOrProvider && Provider.isProvider(alchemyOrProvider)) {
+      superProvider = alchemyOrProvider;
+    }
+    super(privateKey, superProvider);
+
+    // If object passed in is an Alchemy object, just set Alchemy
+    if (alchemyOrProvider && !Provider.isProvider(alchemyOrProvider)) {
+      this.hasAlchemyObject = true;
+      this.alchemy = alchemyOrProvider;
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////
+  // Set of overrides from Signer to handle async provider retrieval.
+  //////////////////////////////////////////////////////////////////
+
+  async getBalance(blockTag?: BlockTag): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.getBalance(blockTag));
+  }
+
+  async getTransactionCount(blockTag?: BlockTag): Promise<number> {
+    return this.getWallet().then(wallet =>
+      wallet.getTransactionCount(blockTag)
+    );
+  }
+
+  // Populates "from" if unspecified, and estimates the gas for the transaction
+  async estimateGas(
+    transaction: Deferrable<TransactionRequest>
+  ): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.estimateGas(transaction));
+  }
+
+  // Populates "from" if unspecified, and calls with the transaction
+  async call(
+    transaction: Deferrable<TransactionRequest>,
+    blockTag?: BlockTag
+  ): Promise<string> {
+    return this.getWallet().then(wallet => wallet.call(transaction, blockTag));
+  }
+
+  // Populates all fields in a transaction, signs it and sends it to the network
+  async sendTransaction(
+    transaction: Deferrable<TransactionRequest>
+  ): Promise<TransactionResponse> {
+    return this.getWallet().then(wallet => wallet.sendTransaction(transaction));
+  }
+
+  async getChainId(): Promise<number> {
+    return this.getWallet().then(wallet => wallet.getChainId());
+  }
+
+  async getGasPrice(): Promise<BigNumber> {
+    return this.getWallet().then(wallet => wallet.getGasPrice());
+  }
+
+  async getFeeData(): Promise<FeeData> {
+    return this.getWallet().then(wallet => wallet.getFeeData());
+  }
+
+  async resolveName(name: string): Promise<string> {
+    return this.getWallet().then(wallet => wallet.resolveName(name));
+  }
+
+  private async getWallet() {
+    if (!this.hasAlchemyObject) {
+      return this.connect(this.provider);
+    }
+    if (!this.alchemy) {
+      throw new Error('No Alchemy object passed into the wallet.');
```

Comment:
This case seems impossible, since `hasAlchemyObject` and `alchemy` are always set together.

Actually, do we need a separate `hasAlchemyObject` field? Seems like we can just check if `this.alchemy` is present.

Reply:
makes sense - cleaned up!

Reply:
It's been removed!